### PR TITLE
osd/PeeringState: fix missed recheck_readable from laggy

### PR DESCRIPTION
**Pull request by huww98**
Previously, the first `pg_lease_ack_t` after becoming laggy would not trigger `recheck_readable`. However, every other ack would trigger it. The logic is inverted, causing unnecessarily long laggy PG state.

Fixes: 3bb8a7210a6 (osd: requeue ops when PG is no longer laggy)
Fixes: https://tracker.ceph.com/issues/53806
Signed-off-by: 胡玮文 huww98@outlook.com

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer (refer to the tracker)
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by huww98** [comment]
Ping. This fix should be trivial. And I've provided the related logs in the tracker. Please help review this.

File: src/osd/PeeringState.cc [hunk 1]
```diff
@@ -1213,14 +1213,14 @@ void PeeringState::proc_lease_ack(int from, const pg_lease_ack_t& a)
...
-	break;
       }
+      break;
     }
   }
   if (was_min) {
     auto old_ru = readable_until;
     recalc_readable_until();
-    if (now < old_ru) {
+    if (now > old_ru) {
```
Comment (jdurgin) [inline src/osd/PeeringState.cc:1223]: to be consistent with the recheck_readable condition this should be >=

**Comment by jdurgin** [comment]
Thanks for the PR, it's a good find, and wonderful that you've got a reproducer as well! Just one minor note.

It'd be even better if you could add an automated test case - one way is using the standalone bash tests (e.g. https://github.com/ceph/ceph/blob/master/qa/standalone/osd/divergent-priors.sh)

**Comment by huww98** [comment]
@jdurgin Sorry for the delay. I missed the notification.

I currently don't have enough time to look into the test. I think the automated test can be hard since the observable impact of this bug is only a slower exit from laggy state. Calculating the timing in the test script can be unstable.

**Comment by rzarzynski** [comment]
@jdurgin: how about retaking a look?

**Comment by huww98** [comment]
Ping @tchaikov @jdurgin . Could you help review this 2 line fix? We would like this to be in the next point release of quincy.

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
hi Sam, as you reviewed #29236 , i am adding you as another reviewer.

**Review by athanatos — APPROVED** [review]
Yep, that looks right.  Good catch!

**Comment by tchaikov** [comment]
- [raodos qa run](https://pulpito.ceph.com/kchai-2022-07-27_14:53:26-rados-wip-kefu-testing-2022-07-27-0857-1-distro-default-smithi/)
- [rerun](https://pulpito.ceph.com/kchai-2022-07-27_14:53:26-rados-wip-kefu-testing-2022-07-27-0857-1-distro-default-smithi/) without #41600

failures tracked by https://tracker.ceph.com/issues/56652

**Comment by ljflores** [comment]
@huww98 we found that this PR has caused a regression. We are reverting the Quincy backport so we can release 17.2.4: https://github.com/ceph/ceph/pull/48104

Please see https://tracker.ceph.com/issues/57546 for more details.

**Pull request merged**